vortex-serde/src/layouts/read/cache.rs

```diff
@@ -26,23 +32,117 @@ impl LayoutMessageCache {
     }
 }
 
+#[derive(Debug)]
+enum LazyDTypeState {
+    Value(DType),
+    Serialized(Bytes, OnceCell<DType>, Projection),
+}
+
+#[derive(Debug)]
+pub struct LazyDeserializedDType {
+    inner: LazyDTypeState,
+}
+
+impl LazyDeserializedDType {
+    pub fn from_bytes(dtype_bytes: Bytes, projection: Projection) -> Self {
+        Self {
+            inner: LazyDTypeState::Serialized(dtype_bytes, OnceCell::new(), projection),
+        }
+    }
+
+    pub fn from_dtype(dtype: DType) -> Self {
+        Self {
+            inner: LazyDTypeState::Value(dtype),
+        }
+    }
+
+    /// Restrict the underlying dtype to selected fields
+    pub fn project(&self, projection: &[Field]) -> VortexResult<Arc<Self>> {
+        match &self.inner {
+            LazyDTypeState::Value(d) => {
+                let DType::Struct(s, n) = d else {
+                    vortex_bail!("Not a struct dtype")
+                };
+                Ok(Arc::new(LazyDeserializedDType::from_dtype(DType::Struct(
+                    s.project(projection)?,
+                    *n,
+                ))))
+            }
+            // TODO(robert): Respect existing projection list, only really an issue for nested structs
+            LazyDTypeState::Serialized(b, ..) => Ok(Arc::new(LazyDeserializedDType::from_bytes(
+                b.clone(),
+                Projection::Flat(projection.to_owned()),
```

Comment:
this projection needs to be projected against the existing one, no? i.e., if I'd previously projected from a struct dtype with fields ["a", "b", "c"] to ["a", "b"], calling project on the latter should "stack" (i.e., I can't subsequently project ["c"]) rather than resetting.

Reply:
I will make it error